tt_metal/llrt/tt_elffile.cpp

```diff
@@ -341,3 +366,311 @@ void ElfFile::Impl::WeakenDataSymbols(std::span<std::string_view const> strong)
         weakener.RewriteSymbols();
     }
 }
+
+void ElfFile::Impl::XIPify() {
+    // In general there can be several lo12 relocs for a hi20
+    // reloc. This is particularly true for lui/{addi,lw,sw,etc}
+    // pairs -- a load and a store might share a single lui, as
+    // the compiler now emits those insns separately. Thus we have
+    // to build a work list and then process it. Furthermore,
+    // although auipc/lo12 pairings are clear because the lo12
+    // part directly points at the auipc, that is not true of
+    // lui/lo12 pairings. We have to use heuristics to locate the
+    // matching relocs and that could get arbitrarily hard. We
+    // presume (a) the compiler doesn't duplicate lui insns, and
+    // (b) the lui preceeds the lo12 in program counter
+    // order. Thus we look for a hi20 reloc matching the symbol at
+    // a lower offset than the lo12 in question. Fortunately we
+    // only need to do this for relocs that need translating, and
+    // those happen to be rare when all data-like sections are in
+    // the data segment (so putting .rodata in text is
+    // problematic). If that proves insufficient here are some
+    // ideas:
+
+    // * Insert fn boundaries from symbols of FNtype -- you'll
+    //   need to tweak the fn address to not cause collisions in
+    //   the reloc map. this might fail with hot/cold block
+    //   splitting.
+
+    // * Construct the CFG by examining R_RISCV_BRANCH
+    //   relocs. Then walk it (backwards) from each lo12 to find
+    //   the reachable hi20. This would be able to deal with
+    //   hot/cold splitting, if one constructed the complete
+    //   section CFG, not as a per-fn entity. One might get away
+    //   with not disasembling to discover ret instructions that
+    //   terminate the CFG.
+
+    struct ComposedReloc {
+        std::vector<Elf32_Rela *> lo_relocs;
+        Elf32_Rela *hi_reloc = nullptr;  // the high part
+
+        ComposedReloc(Elf32_Rela *hi) : hi_reloc(hi) {}
+    };
+
+    enum { ABS, PCREL, HWM };
+    static char const *const r_names[][2] = {
+        {"R_RISCV_HI20", "R_RISCV_LO12"}, {"R_RISCV_PCREL_HI20", "R_RISCV_PCREL_LO12"}};
+
+    auto check_relaxed = [&](Elf32_Rela const &reloc) {
+        // If RELOC is the final reloc, this will
+        // be out of bounds (and probably fail),
+        // but we kind of want that anyway
+        if (ELF32_R_TYPE((&reloc)[1].r_info) != R_RISCV_RELAX)
+            log_debug(tt::LogLLRuntime, "{}: Relocation at {x} is not relaxed", path_, reloc.r_offset);
+    };
+
+    // Place text at 0 for XIP
+    const address_t text_placement = 0;
```

Comment:
We should just kill this variable now -- or as a TODO

Reply:
I'll kill it